django/db/models/sql/query.py

```diff
@@ -815,19 +815,17 @@ def _get_defer_select_mask(self, opts, mask, select_mask=None):
             if filtered_relation := self._filtered_relations.get(field_name):
                 relation = opts.get_field(filtered_relation.relation_name)
                 field_select_mask = select_mask.setdefault((field_name, relation), {})
-                field = relation.field
             else:
-                reverse_rel = opts.get_field(field_name)
+                relation = opts.get_field(field_name)
```

Comment:
Store the related object directly instead of the remote field to be able to differentiate between the two when dealing with self-referential relationship.